### DISCUSSION TITLE
Add a spawn-when-detected action to dynamic window rules

### DISCUSSION THREAD

**Discussion by phuhl** [YaLTeR/niri#273] [Ideas]
For some windows, I'd like to run some code when they appear. This could be done with a spawn-when-detected action.

Example usecase:

When using ediff in Emacs, a new window is spawned, that takes all the user input, but I still want to look at the main Emacs window. The ediff window is full screen sized by default. I can make it small, but I'd like to also shrink the main Emacs window, so both fit on the same screen.

Desired config after ediff window spawned:
![image](https://github.com/YaLTeR/niri/assets/16835755/9dc3bb2a-d4fc-4161-8a7d-1ef53e86b3d3)

Currently: Emacs main window is full screen size, focus is switched to ediff window, so Emacs is actually off screen, so I need to go back to Emacs main window and shrink it, manually.

Usually, one would probably have the Ediff window be floating, but as we don't, yet, have a floating layer, this would be a pretty good solution.

**Comment by YaLTeR** [comment]
Hmm. I'm thinking what would be the best way to solve this problem.

I guess you can make the ediff window small with window rules. What would you `spawn` to also shrink the main emacs window? I guess you'd like it to grow back when ediff is closed? Can it happen that you open ediff while doing something else when you don't want "the window to the left" to change size?

**Reply by phuhl** [reply]
> I guess you can make the ediff window small with window rules. 

Yeah, I am doing that, but in only seems to work with dynamic window rules, not with default-column-width as I have to match by title and I suppose the title is not set from the get go. default-column-width just does not do anything on it.

> What would you spawn to also shrink the main emacs window? 

Just a shell script that moves focus left and shrinks the window, I suppose.

> I guess you'd like it to grow back when ediff is closed?

I mean, that would be epic.

>  Can it happen that you open ediff while doing something else when you don't want "the window to the left" to change size?

No, the window pops up only when I command it to.

**Reply by YaLTeR** [reply]
> I suppose the title is not set from the get go

Might be; it depends on what the client does here.

> I mean, that would be epic.

What I am thinking is that maybe this will be better solved by the event stream IPC which could have an event for a window opening/closing/getting focus, then you could do something in response.
